### ISSUE TITLE
When I add a custom layer that uses stencil tests, some line layers are missing.

### ISSUE THREAD

**Issue by huanglii** [mapbox/mapbox-gl-js#12213]
**mapbox-gl-js version**: 2.10.0

**browser**: Microsoft Edge 版本 105.0.1343.27 (正式版本) (64 位)

### Steps to Trigger Behavior

 1. map.addLayer(testLayer, "aeroway-line"); // some line layers are missing
 2. map.addLayer(testLayer); // line layers are normal

### Link to Demonstration

https://codesandbox.io/s/unruffled-minsky-6e7tpq?file=/index.html

### Actual Behavior
<img width="346" alt="after" src="https://user-images.githubusercontent.com/20068340/188453483-d4a1d829-8439-4bd6-a6f9-7c2e2e2cc278.png">

**Comment by huanglii** [comment]
At the same view, the normal line features are as follows:

<img width="345" alt="before" src="https://user-images.githubusercontent.com/20068340/188528529-b18ff0e7-c1a4-467e-a2c9-b00518fed974.png">

**Comment by mourner** [comment]
Weird... Is this only reproducible in Edge on Windows? Can you check other browsers / OS? Are there any warnings in the console?

**Comment by karimnaaji** [comment]
>@karimnaaji anything we can do on GL JS side so that it's handled automatically?

Yes, we should make sure that the stencil buffer and stencil states are clean between frames. We had a similar issue with retained gl states in https://github.com/mapbox/mapbox-gl-js/issues/10372. The function where we should add state reset is `setCustomLayerDefaults` (which only gets called when we have a custom layer):

https://github.com/mapbox/mapbox-gl-js/blob/2eb020f15069b386e92d261c054f225d75645d47/src/render/painter.js#L901-L916
